### PR TITLE
Suppoert PING, and support raw {tcp,udp,icmp} socket and socket option IP_HDRINCL

### PR DESCRIPTION
**Pull request by skpupil**
support ping: 127.0.0.1 10.0.2.15 and remote(use tap to link with other qemu virtual machine) with IP_HDRINCL or without IP_HDRINCL.
support monitor IP packet from specify iface(10.0.2.15). 
Not support monitor IP packet from all iface, because 0.0.0.0 is not supported in current version(another question).
Not support monitoring IP packets of loopback, Because according to the current implementation, loopback packets are not transmitted through the ip layer.

### CODE REVIEW

**Comment by lrh2000** [comment]
Thanks for the PR! Really appreciate on your work to make all these things work. I'll do my best to find some time to review this PR in the next few days.

However, I'm afraid there will be conflicts due to the upcoming backlog optimization, which will probably be prioritized according to our current schedule. In the meantime, this PR is just too big, so I'm afraid it will take some time to complete the review process and get us to agree on the implementation details. I apologize in advance and hope for your understanding.

**Comment by skpupil** [comment]
> Thanks for the PR! Really appreciate on your work to make all these things work. I'll do my best to find some time to review this PR in the next few days.
> 
> However, I'm afraid there will be conflicts due to the upcoming backlog optimization, which will probably be prioritized according to our current schedule. In the meantime, this PR is just too big, so I'm afraid it will take some time to complete the review process and get us to agree on the implementation details. I apologize in advance and hope for your understanding.

okok. No problem. Feel free and take your time.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        };
+
+        let bound_raw = match unbound_raw.bind(addr, can_reuse) {
+            Ok(bound_raw) => bound_raw,
+            Err((err, unbound_raw)) => return Err((err, Inner::Unbound(unbound_raw))),
+        };
+        Ok(bound_raw)
+    }
+
+    fn bind_to_ephemeral_endpoint(
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:91]: I don't think raw sockets need an ephemeral port. I believe that raw sockets have nothing to do with the port.

```rust
        // Raw socket is no need to release port
```
As you said, raw sockets will not free the port, so it should not be bound to a port either, otherwise the port can never be freed.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    fn get_option(&self, option: &mut dyn SocketOption) -> Result<()> {
+        match_sock_option_mut!(option, {
+            socket_errors: SocketError => {
+                self.options.write().socket.get_and_clear_sock_errors(socket_errors);
+                return Ok(());
+            },
+            _ => ()
+        });
+        let _ = self.options.read().socket.get_option(option);
+        self.options.read().ip.get_option(option)
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:398]: Please add blank lines between set/get options from different levels. This will make the code a bit clearer.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+
+    fn get_option(&self, option: &mut dyn SocketOption) -> Result<()> {
+        match_sock_option_mut!(option, {
+            socket_errors: SocketError => {
+                self.options.write().socket.get_and_clear_sock_errors(socket_errors);
+                return Ok(());
+            },
+            _ => ()
+        });
+        let _ = self.options.read().socket.get_option(option);
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:397]: If this succeeds, you won't need to call `ip.get_option`, right?

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        self.options.read().ip.get_option(option)
+    }
+
+    fn set_option(&self, option: &dyn SocketOption) -> Result<()> {
+        if self.options.write().socket.set_option(option).is_ok() {
+            return Ok(());
+        }
+        if self.options.write().ip.set_option(option).is_ok() {
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:405]: It is possible to just return whatever `self.options.write().ip.set_option(option)` returns without checking the return value, right?

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+                return Ok(());
+            },
+            _ => ()
+        });
+        let _ = self.options.read().socket.get_option(option);
+        self.options.read().ip.get_option(option)
+    }
+
+    fn set_option(&self, option: &dyn SocketOption) -> Result<()> {
+        if self.options.write().socket.set_option(option).is_ok() {
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:402]: Write `let options = self.options.write()` to avoid taking the write lock multiple times.

Same for `get_option`, where taking the read lock multiple times is also unnecessary.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+
+        // This is a temporary solution. Bind to a specific iface: iface[0]
+        // TODO: Monitor multiple ifaces
+        let bind_iface_addr = IpEndpoint::new(IpAddress::v4(10, 0, 2, 15), 0);
+        self.try_bind_ephemeral(&bind_iface_addr)?;
+
+        let (received_bytes, addr) = self.recv(writer, flags)?;
+        // TODO: Receive control message
+        let IpAddress::Ipv4(ip_addr) = addr;
+        let message_header = MessageHeader::new(Some(SocketAddr::IPv4(ip_addr, 0)), None);
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:385]: Add regression tests to ensure this is the correct Linux behavior.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+        Ok((received_bytes, message_header))
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:386]: Add blank lines to split unrelated code blocks. I think they're required:
 - before `TODO: Receive control message`
 - after `TODO: Receive control message`
 - after `let message_header = ...`

File: kernel/src/net/socket/ip/raw/unbound.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+pub struct UnBoundRaw {
+    unbound_socket: Box<UnboundRawSocket>,
+    ip_protocol: IpProtocol,
+}
+
+impl UnBoundRaw {
+    pub fn new(observer: Weak<dyn SocketEventObserver>, protocol: Protocol) -> Self {
+        let ip_protocol = match protocol {
+            //support more?
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/unbound.rs:23]: Again,
> Please make this sentence as a correct English sentence (e.g., the first letter should be capitalized) and insert a whitespace after //.

File: test/apps/network/raw_tcp_server.c [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	// Create a raw socket to listen for incoming TCP packets
+	sock = socket(AF_INET, SOCK_RAW, IPPROTO_TCP);
+	if (sock < 0) {
+		perror("Socket creation failed");
+		return EXIT_FAILURE;
+	}
+
+	printf("Server is listening for raw TCP packets...\n");
+
+	// Infinite loop to receive and print incoming packets
```
Comment (lrh2000) [inline test/apps/network/raw_tcp_server.c:28]: Where's the infinite loop?

**Review by lrh2000 — COMMENTED** [review]
Thanks for your work to support raw sockets and support the PING program!
 - For the implementation in `aster-bigtcp`, I would like to make sure that packets sent from the raw sockets are properly handled by the network stack, and that packets passing through the network stack can be properly captured by the network stack.
    -  Unfortunately, we don't have a unit test framework to ensure this, but perhaps we can add more regression tests to test the end-to-end behavior.
 - For the implementation in `kernel`, it looks like you're mostly copying the logic from UDP sockets. But I think raw sockets behave in a way that is significantly different from UDP sockets (e.g., `bind()/connect()` system calls), so I think we need more testing to understand the Linux behavior. 
   - I suggest you check the tests in `test/apps/network/tcp_err.c` and do similar things to add a `test/apps/network/raw_err.c`.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+        let socket = SocketOptionSet::new_raw();
+        let ip = IpSocketOptionSet::new_raw();
+        OptionSet { socket, ip }
+    }
+}
+
+pub struct RawDGramSocket {
+    options: RwLock<OptionSet>,
+    inner: RwLock<Takeable<Inner>, LocalIrqDisabled>,
+    nonblocking: AtomicBool,
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:59]: I know this code is copied from UDP sockets and they name the variable as `nonblocking`. But please name this variable as `is_nonblocking` is new code, which is a better name since `is_xxx` makes it clear that it's a boolean variable.

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+            }
+        })
+    }
+
+    pub fn is_nonblocking(&self) -> bool {
+        self.nonblocking.load(Ordering::SeqCst)
+    }
+
+    pub fn set_nonblocking(&self, nonblocking: bool) {
+        self.nonblocking.store(nonblocking, Ordering::SeqCst);
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:122]: `is_nonblocking` should be accessed with `Ordering::Relaxed` unless you want to synchronize something.

File: kernel/libs/aster-bigtcp/src/iface/poll.rs [hunk 12]
```diff
@@ -90,7 +95,10 @@ impl<E> PollContext<'_, E> {
             );
         }
 
+        self.raw_socket_filter(&IpRepr::Ipv4(repr), pkt.payload());
```
[the remark this answers is not in the document]
Comment (skpupil) [reply kernel/libs/aster-bigtcp/src/iface/poll.rs:98]: Do you mean to filter the tcp or udp packets sent to 127.0.0.1 at the ip layer?  Thanks.

[the remark this answers is not in the document]
Comment (skpupil) [reply kernel/libs/aster-bigtcp/src/iface/poll.rs:98]: > Do you mean to filter the tcp or udp packets sent to 127.0.0.1 at the ip layer? Thanks.

Do you mean to filter the tcp or udp packets sent to 127.0.0.1 at the ip layer? Thanks.

[the remark this answers is not in the document]
Comment (skpupil) [reply kernel/libs/aster-bigtcp/src/iface/poll.rs:98]: > Could you provide equivalent versions of this method for TCP/UDP and call it in `dispatch_tcp`/`dispatch_udp`/`process_tcp_until_outgoing` to make the raw sockets work for locally generated and locally consumed packets? I think this is possible, please correct me if I'm wrong.

Do you mean to filter the tcp or udp packets sent to 127.0.0.1 at the ip layer?

[the remark this answers is not in the document]
Comment (lrh2000) [reply kernel/libs/aster-bigtcp/src/iface/poll.rs:98]: > Do you mean to filter the tcp or udp packets sent to 127.0.0.1 at the ip layer?

Yes, I mean we can support using raw sockets to capture packets sent to `127.0.0.1` (assuming the raw socket is bound to the loopback iface). I think it's doable in the current codebase, just by adding some variants of this `raw_socket_filter` method.

File: kernel/libs/aster-bigtcp/src/socket/bound.rs [hunk 13]
```diff
@@ -40,25 +43,27 @@ pub trait AnySocket {
...
-/// Common states shared by [`BoundTcpSocketInner`] and [`BoundUdpSocketInner`].
+/// Common states shared by [`BoundTcpSocketInner`] , [`BoundUdpSocketInner`] and [`BoundRawSocketInner`].
 pub struct BoundSocketInner<T, E> {
     iface: Arc<dyn Iface<E>>,
     port: u16,
     socket: T,
     observer: RwLock<Weak<dyn SocketEventObserver>, LocalIrqDisabled>,
     events: AtomicU8,
     next_poll_at_ms: AtomicU64,
+    ip_protocol: IpProtocol,
```
[the remark this answers is not in the document]
Comment (skpupil) [reply kernel/libs/aster-bigtcp/src/socket/bound.rs:56]: The protocol field is used to filter out network packets suitable for the socket. This field should be placed in the socketBg structure. However, at this time, this field is redundant for tcp and udp, as you said, and the port field is redundant for raw sockets. Using one structure to abstract all socket types is always not perfect, and there are more or less redundant fields. Maybe we should just accept this fact and allow these redundant fields to exist? 
If there is a better way, please tell me, thanks.

[the remark this answers is not in the document]
Comment (skpupil) [reply kernel/libs/aster-bigtcp/src/socket/bound.rs:56]: > The protocol field is used to filter out network packets suitable for the socket. This field should be placed in the socketBg structure. However, at this time, as you said, this field is redundant for tcp and udp, and the port field is redundant for raw sockets. Using one structure to abstract all socket types is always not perfect, and there are more or less redundant fields. Maybe we should just accept this fact and allow these redundant fields to exist?
> If there is a better way, please tell me, thanks.


The raw socket structure in smoltcp has a protocol field, which temporarily solves my problem. However, the port field is redundant for raw sockets, which may be acceptable？

[the remark this answers is not in the document]
Comment (lrh2000) [reply kernel/libs/aster-bigtcp/src/socket/bound.rs:56]: > However, the port field is redundant for raw sockets, which may be acceptable？

I admit it's a pain to deal with this problem.

In fact, even if we don't consider raw sockets, TCP/UDP sockets also have different meanings in the port field.
 - In other words, TCP ports and UDP ports should be assigned independently, i.e., there can be a TCP socket and a UDP socket that have the same port assignment, but there should be no conflict because they are different protocols, but this is not correctly implemented in Asterinas for now.

For raw sockets, it has no port assignment, so you may want to use something like:
```rust
struct BoundPort {
    iface: Arc<dyn Iface>,
    port: Option<NonZeroU16>,
}
```
Then, just let `port` be `None`.

We can consider the refactoring of `BoundPort` as future work.

**Comment by skpupil** [comment]
@lrh2000 While reading the code, I kept thinking, do we really need such abstraction, with three socket types in total, and not more. We are trying to abstract TCP, UDP and raw sockets so that they can fit into a common structure, but this does not seem to bring significant benefits. On the contrary, it is very inflexible and often needs to be refactored repeatedly because of the need to increase functionality. Perhaps, for software as complex as an operating system, less abstraction and more process-oriented code is more appropriate?

**Comment by skpupil** [comment]
> @lrh2000 @tatetian While reading the code, I kept thinking, do we really need such abstraction, with three socket types in total, and not more. We are trying to abstract TCP, UDP and raw sockets so that they can fit into a common structure, As mentioned above, socketBg is still difficult to perfectly abstract various socket types. and this does not seem to bring significant benefits. On the contrary, it is very inflexible and often needs to be refactored repeatedly because of the need to increase functionality. At present, when the amount of code is not large, such frequent refactoring is required. Will such refactoring become more and more difficult when the amount of code reaches a large scale in the future? In addition, this method makes the entire software affected by a single move, and each refactoring can only be performed by one person, which seems to affect efficiency. Perhaps, for software as complex as an operating system, less abstraction and more process-oriented code is more appropriate?

What we are doing is using struct to abstract different types of objects, which is perhaps not the recommended style? I understand that the recommendation in Rust is to use traits for behavioral abstraction. This may be the reason for the [pain](https://github.com/asterinas/asterinas/pull/1660#discussion_r1897038394).

**Comment by lrh2000** [comment]
> @lrh2000 @tatetian While reading the code, I kept thinking, do we really need such abstraction, with three socket types in total, and not more. We are trying to abstract TCP, UDP and raw sockets so that they can fit into a common structure, As mentioned above, socketBg is still difficult to perfectly abstract various socket types. and this does not seem to bring significant benefits.

Historically, we have had a single type, `AnyBoundSocket`, which represents "any" (i.e., TCP/UDP) bound socket. This type implements the "common" logic for all sockets. Later we will refactor the code to split it into `BoundTcpSocket`/`BoundUdpSocket` and try to find a way to abstract the common logic _and_ allow each type to have protocol-dependent logic.

So we're currently in the process of _breaking things_ rather than trying to _abstract things_. The abstraction comes because it's there at a very, very early stage of the Asterinas kernel.

But do we want to remove abstraction altogether? I think the answer is still no, because it avoids some redundant code. Having redundant code is also annoying, and makes development difficult in another way: it is easy to forget to change redundant code, but maybe everything just compiles and no one notices.

I don't think the current abstraction is perfect. If you have a concrete design to improve the abstraction, we can make things actionable by writing an RFC and discussing it.

> On the contrary, it is very inflexible and often needs to be refactored repeatedly because of the need to increase functionality. At present, when the amount of code is not large, such frequent refactoring is required. Will such refactoring become more and more difficult when the amount of code reaches a large scale in the future? In addition, this method makes the entire software affected by a single move, and each refactoring can only be performed by one person, which seems to affect efficiency.

It is _not_ the abstraction layer that causes frequent refactoring. The main reason we need frequent refactoring at this point is that we're changing the code from smoltcp-style to linux-style. This has to be done step by step, and many steps are not trivial, so it involves a larger number of code changes.

In other words, I think that even if we do not abstract anything, you need to adjust this PR to make sure that all your code conforms to the mainline code as well.
 - For example, if we don't abstract anything, you can just copy some code from TCP/UDP sockets to raw sockets, then after you do the refactoring, the code pattern of TCP/UDP sockets is different, so you have to adjust your code to make sure it still matches.
 - This is _even worse_, because it increases the review overhead (the reviewer has to make sure that all the legacy code you copied is updated properly).

**Comment by lrh2000** [comment]
> What we are doing is using struct to abstract different types of objects, which is perhaps not the recommended style? I understand that the recommendation in Rust is to use traits for behavioral abstraction. This may be the reason for the [pain](https://github.com/asterinas/asterinas/pull/1660#discussion_r1897038394).

Again, I don't think the current abstraction is perfect. If you have a better design, we can discuss it.

**Comment by skpupil** [comment]
> > @lrh2000 @tatetian While reading the code, I kept thinking, do we really need such abstraction, with three socket types in total, and not more. We are trying to abstract TCP, UDP and raw sockets so that they can fit into a common structure, As mentioned above, socketBg is still difficult to perfectly abstract various socket types. and this does not seem to bring significant benefits.
> 
> Historically, we have had a single type, `AnyBoundSocket`, which represents "any" (i.e., TCP/UDP) bound socket. This type implements the "common" logic for all sockets. Later we will refactor the code to split it into `BoundTcpSocket`/`BoundUdpSocket` and try to find a way to abstract the common logic _and_ allow each type to have protocol-dependent logic.
> 
> So we're currently in the process of _breaking things_ rather than trying to _abstract things_. The abstraction comes because it's there at a very, very early stage of the Asterinas kernel.
> 
> But do we want to remove abstraction altogether? I think the answer is still no, because it avoids some redundant code. Having redundant code is also annoying, and makes development difficult in another way: it is easy to forget to change redundant code, but maybe everything just compiles and no one notices.
> 
> I don't think the current abstraction is perfect. If you have a concrete design to improve the abstraction, we can make things actionable by writing an RFC and discussing it.
> 
> > On the contrary, it is very inflexible and often needs to be refactored repeatedly because of the need to increase functionality. At present, when the amount of code is not large, such frequent refactoring is required. Will such refactoring become more and more difficult when the amount of code reaches a large scale in the future? In addition, this method makes the entire software affected by a single move, and each refactoring can only be performed by one person, which seems to affect efficiency.
> 
> It is _not_ the abstraction layer that causes frequent refactoring. The main reason we need frequent refactoring at this point is that we're changing the code from smoltcp-style to linux-style. This has to be done step by step, and many steps are not trivial, so it involves a larger number of code changes.
> 
> In other words, I think that even if we do not abstract anything, you need to adjust this PR to make sure that all your code conforms to the mainline code as well.
> 
> * For example, if we don't abstract anything, you can just copy some code from TCP/UDP sockets to raw sockets, then after you do the refactoring, the code pattern of TCP/UDP sockets is different, so you have to adjust your code to make sure it still matches.
> * This is _even worse_, because it increases the review overhead (the reviewer has to make sure that all the legacy code you copied is updated properly).

Got it. For now, I won't consider bypassing some abstractions, such as `socketBg`, and let's design it after this pr in the future.

**Comment by skpupil** [comment]
I have resubmitted this PR, please take the time to review it for me. Thank you. @lrh2000

File: kernel/src/net/socket/ip/raw/unbound.rs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+use aster_bigtcp::{
+    iface::BoundPort,
+    socket::RawSocket,
+    wire::{IpAddress, IpProtocol},
+};
+
+use super::{bound::BoundRaw, RawSocketObserver};
+use crate::{events::IoEvents, net::socket::ip::common::get_ephemeral_iface, prelude::*};
+
+pub struct UnBoundRaw {
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/unbound.rs:12]:
```suggestion
pub struct UnboundRaw {
```

File: kernel/src/net/socket/ip/raw/mod.rs [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+        bound_raw.set_remote_endpoint(&endpoint);
+
+        Ok(())
+    }
+
+    fn addr(&self) -> Result<SocketAddr> {
+        let inner = self.inner.read();
+        match inner.as_ref() {
+            Inner::Unbound(_) => Ok(UNSPECIFIED_LOCAL_ENDPOINT.into()),
+            Inner::Bound(bound_raw) => Ok(bound_raw.local_endpoint().into()),
```
Comment (lrh2000) [inline kernel/src/net/socket/ip/raw/mod.rs:327]: This implemetantion is unlikely to be correct because the result "port" should represent the protocol:
```python
Python 3.13.1 (main, Dec  4 2024, 18:05:56) [GCC 14.2.1 20240910] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_TCP)
>>> s.getsockname()
('0.0.0.0', 6)
>>> socket.IPPROTO_TCP
6
```

File: test/apps/network/raw_err.c [hunk 16]
```diff
@@ -0,0 +1,640 @@
...
+	struct sockaddr_in saddr = { .sin_port = 0xbeef };
+	struct sockaddr *psaddr = (struct sockaddr *)&saddr;
+	socklen_t addrlen = 0;
+
+	TEST_RES(getsockname(sk_unbound, psaddr, &addrlen),
+		 addrlen == sizeof(saddr) && saddr.sin_port == 0xbeef);
+
+	TEST_RES(getsockname(sk_unbound, psaddr, &addrlen),
+		 addrlen == sizeof(saddr) &&
+			 saddr.sin_addr.s_addr == inet_addr("0.0.0.0"));
```
Comment (lrh2000) [inline test/apps/network/raw_err.c:109]: Check the port in the regression tests. Don't ignore it.

File: test/apps/network/raw_err.c [hunk 17]
```diff
@@ -0,0 +1,640 @@
...
+
+	TEST_RES(getsockname(sk_bound, psaddr, &addrlen),
+		 addrlen == sizeof(saddr) &&
+			 saddr.sin_addr.s_addr == inet_addr("127.0.0.1"));
+
+	TEST_RES(getsockname(sk_connected, psaddr, &addrlen),
+		 addrlen == sizeof(saddr) &&
+			 saddr.sin_addr.s_addr == inet_addr("127.0.0.1") &&
+			 saddr.sin_port != C_PORT); //saddr.sin_port != C_PORT);
```
Comment (lrh2000) [inline test/apps/network/raw_err.c:118]: Remove the commented code.

File: kernel/libs/aster-bigtcp/src/iface/poll.rs [hunk 18]
```diff
@@ -509,4 +632,75 @@ impl<E: Ext> PollContext<'_, E> {
...
+            // We set `did_something` even if no packets are actually generated. This is because a
+            // timer can expire, but no packets are actually generated.
+            did_something = true;
+
+            let mut deferred = None;
+
+            socket.dispatch(self.iface_cx, |cx, ip_repr, ip_payload| {
+                let this = PollContext::new(cx, self.sockets, self.new_tcp_conns);
+                if ip_repr.dst_addr().is_broadcast() || !this.is_unicast_local(ip_repr.dst_addr()) {
+                    dispatch_phy(
```
Comment (lrh2000) [inline kernel/libs/aster-bigtcp/src/iface/poll.rs:657]: If it's not a broadcast packet, we should make raw packets capture such packets. (Broadcast packets will be captured in `parse_and_process_ipv4` anyway).

File: kernel/libs/aster-bigtcp/src/iface/poll.rs [hunk 19]
```diff
@@ -337,9 +421,14 @@ impl<E: Ext> PollContext<'_, E> {
...
+            return did_something_udp;
+        };
+
+        let (did_something_raw, _tx_token) = self.dispatch_raw(tx_token, dispatch_phy);
 
-        did_something_tcp || did_something_udp
+        did_something_tcp || did_something_udp || did_something_raw
     }
 
     fn dispatch_tcp<T, Q>(&mut self, tx_token: T, dispatch_phy: &mut Q) -> (bool, Option<T>)
```
Comment (lrh2000) [inline kernel/libs/aster-bigtcp/src/iface/poll.rs:434]: For TCP sockets, `dispatch` also processes generated responses:
https://github.com/asterinas/asterinas/blob/d958ea3153935315d630b7a80324fc4cf8ff8a7b/kernel/libs/aster-bigtcp/src/socket/bound.rs#L991-L999

We need to ensure that each packet can be captured by raw sockets once and only once, so we need to make some adjustments to this logic.

**Comment by tatetian** [comment]
I am closing this PR as it has been inactive for too long. The author is still encouraged to resume this line of work.
